src/collection.ts

```diff
@@ -1,33 +1,188 @@
+import GitDB from './gitdb';
+import fse from 'fs-extra';
 interface Record {
-  id: string
+  id: string;
 }
 
 interface Filter {
-
+  id?: string;
 }
 
 type SetCallback<T> = (record: T) => T;
-
 type UpdateOrCallback<T, K extends T> = T | SetCallback<K>;
 
 /**
  * TODO add immutablejs to prevent array and object mutation
  */
-export default class Collection extends Array {
-  public insert<T>(data: T): Promise<T & Record> {
-    // TODO insert logic
-    return Promise.resolve({...data, id: String(Math.random())});
+export default class Collection<T extends Record> {
+  private db: GitDB;
+  private name: string;
+  private data: any[];
+  [n: number]: T;
+  get length() {
+    return this.data.length;
+  }
+
+  private constructor(gitDb: GitDB, name: string) {
+    this.db = gitDb;
+    this.name = name;
+    this.data = [];
+  }
+
+  private static handler: ProxyHandler<Collection<any>> = {
+    get: (target, prop, receiver) => {
+      const propAsNumber = Number(prop);
+      if (isNaN(propAsNumber)) {
+        /** If it is method or property of collection, it's ok, else throws error  */
+        if (
+          /** @see https://eslint.org/docs/rules/no-prototype-builtins */
+          Object.prototype.hasOwnProperty.call(Collection.prototype, prop) ||
+          Object.prototype.hasOwnProperty.call(target, prop)
+        ) {
+          // Element implicitly has an 'any' type because index expression is not of type 'number'.
+          // But we want to access methods & private properties, not a collection[NUMBER] value
+          //@ts-ignore
+          return target[prop];
+        } else {
+          throw `Property [${String(prop)}] of collection [${
+            target.name
+          }] is inaccessible`;
+        }
+      } else {
+        return target.data[propAsNumber];
+      }
+    }
+  };
+
+  public static createCollection(name: string, gitDB: GitDB): Collection<any> {
+    const collection = new Collection(gitDB, name);
+    const collectionProxy = new Proxy(collection, Collection.handler);
+    return collectionProxy;
   }
 
-  public update<T, K extends T>(filter: Filter, dataOrCallback: UpdateOrCallback<T, K>): Promise<K> {
-    // TODO insert logic
+  public async insert(documentData: T, writeToDisk = true): Promise<T> {
+    const newDocument = { id: String(Math.random()), ...documentData };
+    this.data.push(newDocument);
+    const filePath = `${this.db.config.dbDir}/${this.name}/${newDocument.id}.json`;
+    if (writeToDisk) {
+      fse.outputJson(filePath, newDocument);
+    }
+    return newDocument;
+  }
+
+  public async update<K extends T>(
+    filter: Filter,
+    dataOrCallback: UpdateOrCallback<T, K>
+  ): Promise<K> {
+    const { id } = filter;
+    if (typeof dataOrCallback !== 'function') {
+      if (id) {
+        const documentIndex = this.data.findIndex((e) => (e.id = id));
+        this.data.splice(documentIndex, 1);
+        this.data = [...this.data, dataOrCallback];
```

Comment:
we need to be checking if dataOrCallback is a function and call it to get `newData`:
```js
if (typeof dataOrCallback  == 'function') {
  this.data[documentIndex] = [
    ...this.data.slice(0, documentIndex),  
    dataOrCallback(this.data[documentIndex]),
    ...this.data.slice(documentIndex + 1),  
    ]
}
```